### PR TITLE
Eliminate c10::guts::to_string

### PR DESCRIPTION
**Pull request by cyyever**
This PR replace c10::guts::to_string with std::to_string. The major part of changes is using void* as optimizer state key since string is used only for serialization and using pointers as hashing keys is more efficient than a string. 
Some other guts functions in the affected source files are also replaced.

### CODE REVIEW

**Comment by Skylion007** [comment]
While we are doing this, maybe be we should switch it to fmtlib's to_string as that implementation should be more efficient?

**Comment by cyyever** [comment]
> While we are doing this, maybe be we should switch it to fmtlib's to_string as that implementation should be more efficient?

Using fmtlib is planed as another PR, we should focus on cleaning up old functions here.

File: test/cpp/api/serialize.cpp [hunk 1]
```diff
@@ -218,15 +218,15 @@ TEST(SerializeTest, KeysFunc) {
   torch::serialize::OutputArchive output_archive;
   for (const auto i : c10::irange(3)) {
     output_archive.write(
-        "element/" + c10::to_string(i), c10::IValue(static_cast<int64_t>(i)));
+        "element/" + std::to_string(i), c10::IValue(static_cast<int64_t>(i)));
```
Comment (Skylion007) [inline test/cpp/api/serialize.cpp:221]: There is also a fmt::to_string, but it's not in the standard fmt header

Comment (cyyever) [reply test/cpp/api/serialize.cpp:221]: I would use fmt::format here

Comment (Skylion007) [reply test/cpp/api/serialize.cpp:221]: Ah, makes sense.

**Review by Skylion007 — APPROVED** [review]

**Comment by huydhn** [comment]
@pytorchbot revert -m 'Sorry for reverting this, but this is needed to keep trunk green after https://github.com/pytorch/pytorch/pull/108479 was reverted.  Both will need to be relanded' -c ghfirst

https://hud.pytorch.org/pytorch/pytorch/commit/9f71a4ebd462b95e1b45df4a89f00ac83b9d77ac fails because `to_string` has been removed from `c10::guts`

**Pull request closed without merging**